runtime/client_http_request.go

```diff
@@ -203,21 +212,80 @@ func (req *ClientHTTPRequest) Do() (*ClientHTTPResponse, error) {
 		/* coverage ignore next line */
 		return nil, err
 	}
+	var retryCount int64 = 1
+	var res *http.Response
+
+	//when timeoutAndRetryOptions per request is not configured, use default client level timeout
+	if req.timeoutAndRetryOptions == nil || req.timeoutAndRetryOptions.MaxAttempts == 0 {
+		res, err = req.client.Client.Do(req.httpReq.WithContext(ctx))
+	} else {
+		res, retryCount, err = req.executeDoWithRetry(ctx) //new code for retry and timeout per ep level
+	}
 
-	res, err := req.client.Client.Do(req.httpReq.WithContext(ctx))
 	span.Finish()
+
 	if err != nil {
-		req.ContextLogger.ErrorZ(req.ctx, "Could not make outbound request", zap.Error(err))
-		return nil, err
+		req.ContextLogger.ErrorZ(req.ctx, fmt.Sprintf("Could not make http outbound %s.%s request",
+			req.ClientID, req.MethodName), zap.Error(err))
+		return nil, errors.Wrapf(err, "errors while making outbound %s.%s request", req.ClientID, req.MethodName)
 	}
 
 	// emit metrics
-	req.Metrics.IncCounter(req.ctx, clientRequest, 1)
+	req.Metrics.IncCounter(req.ctx, clientRequest, retryCount)
 
 	req.res.setRawHTTPResponse(res)
 	return req.res, nil
 }
 
+//executeDoWithRetry will execute executeDo with retries
+func (req *ClientHTTPRequest) executeDoWithRetry(ctx context.Context) (*http.Response, int64, error) {
+	var err error
+	var res *http.Response
+	var retryCount int64 = 0
+
+	for i := 0; i < req.timeoutAndRetryOptions.MaxAttempts; i++ {
+		retryCount++
+		res, err = req.executeDo(ctx)
+
+		if err == nil {
+			return res, retryCount, nil
+		}
+
+		var shouldRetry = false
+		// if attempts are pending, wait for backoff duration before next attempt
+		if i+1 < req.timeoutAndRetryOptions.MaxAttempts {
+			shouldRetry = req.client.CheckRetry(ctx, req.timeoutAndRetryOptions, res, err)
+		}
+
+		req.ContextLogger.GetLogger().Info("errors while making http outbound request",
```

Comment:
should be an error log?

Reply:
also why did we expose the underlying logger instead of using the context logger directly?

Reply:
Have updated it as **Warn** as retries could be pending which says it's not the final error. 
Why not contextLogger? Each attempt may fail with diff reason and may override the message in contextLogger, what we want is individual logs for each retry. Other approach is to fix contextLogger to have same field with multiple values, like msg1, msg2..